File: app/src/main/java/com/elementary/tasks/core/app_widgets/calendar/CalendarWidget.java
package com.elementary.tasks.core.app_widgets.calendar;

import android.annotation.TargetApi;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import androidx.annotation.NonNull;
import android.text.format.DateUtils;
import android.widget.RemoteViews;

import com.elementary.tasks.R;
import com.elementary.tasks.core.app_widgets.WidgetUtils;
import com.elementary.tasks.core.app_widgets.voice_control.VoiceWidgetDialog;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.creators.CreateReminderActivity;
import com.elementary.tasks.navigation.MainActivity;

import java.util.Calendar;
import java.util.Formatter;
import java.util.GregorianCalendar;
import java.util.Locale;

/**
 * Copyright 2015 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class CalendarWidget extends AppWidgetProvider {

    @Override
    public void onReceive(@NonNull Context context, @NonNull Intent intent) {
        super.onReceive(context, intent);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        SharedPreferences sp = context.getSharedPreferences(
                CalendarWidgetConfig.CALENDAR_WIDGET_PREF, Context.MODE_PRIVATE);
        for (int i : appWidgetIds) {
            updateWidget(context, appWidgetManager, sp, i);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                Bundle options = appWidgetManager.getAppWidgetOptions(i);
                onAppWidgetOptionsChanged(context, appWidgetManager, i,
                        options);
            }
        }
        super.onUpdate(context, appWidgetManager, appWidgetIds);
    }

    public static void updateWidget(Context context, AppWidgetManager appWidgetManager,
                                    SharedPreferences sp, int widgetID) {
        Calendar cal = new GregorianCalendar();
        int month = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_MONTH + widgetID, 0);
        int year = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_YEAR + widgetID, 0);
        cal.set(Calendar.MONTH, month);
        cal.set(Calendar.YEAR, year);
        StringBuilder monthYearStringBuilder = new StringBuilder(50);
        Formatter monthYearFormatter = new Formatter(
                monthYearStringBuilder, Locale.getDefault());
        int monthYearFlag = DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_NO_MONTH_DAY | DateUtils.FORMAT_SHOW_YEAR;
        String date = DateUtils.formatDateRange(context,
                monthYearFormatter, cal.getTimeInMillis(), cal.getTimeInMillis(), monthYearFlag)
                .toString().toUpperCase();

        int theme = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_THEME + widgetID, 0);
        CalendarTheme item = CalendarTheme.getThemes(context).get(theme);

        RemoteViews rv = new RemoteViews(context.getPackageName(), R.layout.calendar_widget_layout);
        rv.setTextViewText(R.id.currentDate, date);
        rv.setTextColor(R.id.currentDate, item.getTitleColor());

        rv.setInt(R.id.weekdayGrid, "setBackgroundResource", item.getWidgetBgColor());
        rv.setInt(R.id.header, "setBackgroundResource", item.getHeaderColor());
        rv.setInt(R.id.monthGrid, "setBackgroundResource", item.getBorderColor());

        WidgetUtils.setIcon(context, rv, item.getIconPlus(), R.id.plusButton);
        WidgetUtils.setIcon(context, rv, item.getIconVoice(), R.id.voiceButton);
        WidgetUtils.setIcon(context, rv, item.getIconSettings(), R.id.settingsButton);
        WidgetUtils.setIcon(context, rv, item.getRightArrow(), R.id.nextMonth);
        WidgetUtils.setIcon(context, rv, item.getLeftArrow(), R.id.prevMonth);

        Intent weekdayAdapter = new Intent(context, CalendarWeekdayService.class);
        weekdayAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetID);
        rv.setRemoteAdapter(R.id.weekdayGrid, weekdayAdapter);

        Intent startActivityIntent = new Intent(context, MainActivity.class);
        startActivityIntent.putExtra(Constants.INTENT_POSITION, R.id.nav_calendar);
        PendingIntent startActivityPendingIntent = PendingIntent.getActivity(context, 0,
                startActivityIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        rv.setPendingIntentTemplate(R.id.monthGrid, startActivityPendingIntent);

        Intent monthAdapter = new Intent(context, CalendarMonthService.class);
        monthAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetID);
        rv.setRemoteAdapter(R.id.monthGrid, monthAdapter);

        Intent configIntent = new Intent(context, CreateReminderActivity.class);
        PendingIntent configPendingIntent = PendingIntent.getActivity(context, 0, configIntent, 0);
        rv.setOnClickPendingIntent(R.id.plusButton, configPendingIntent);

        configIntent = new Intent(context, VoiceWidgetDialog.class);
        configPendingIntent = PendingIntent.getActivity(context, 0, configIntent, 0);
        rv.setOnClickPendingIntent(R.id.voiceButton, configPendingIntent);

        configIntent = new Intent(context, CalendarWidgetConfig.class);
        configIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetID);
        configPendingIntent = PendingIntent.getActivity(context, 0, configIntent, 0);
        rv.setOnClickPendingIntent(R.id.settingsButton, configPendingIntent);

        Intent serviceIntent = new Intent(context, CalendarUpdateService.class);
        serviceIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetID);
        serviceIntent.putExtra("actionPlus", 2);
        PendingIntent servicePendingIntent =
                PendingIntent.getService(context, 0, serviceIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        rv.setOnClickPendingIntent(R.id.nextMonth, servicePendingIntent);

        serviceIntent = new Intent(context, CalendarUpdateMinusService.class);
        serviceIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetID);
        serviceIntent.putExtra("actionMinus", 1);
        servicePendingIntent =
                PendingIntent.getService(context, 0, serviceIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        rv.setOnClickPendingIntent(R.id.prevMonth, servicePendingIntent);

        appWidgetManager.updateAppWidget(widgetID, rv);
        appWidgetManager.notifyAppWidgetViewDataChanged(widgetID, R.id.weekdayGrid);
        appWidgetManager.notifyAppWidgetViewDataChanged(widgetID, R.id.monthGrid);
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        SharedPreferences.Editor editor = context.getSharedPreferences(
                CalendarWidgetConfig.CALENDAR_WIDGET_PREF, Context.MODE_PRIVATE).edit();
        for (int widgetID : appWidgetIds) {
            editor.remove(CalendarWidgetConfig.CALENDAR_WIDGET_THEME + widgetID);
            editor.remove(CalendarWidgetConfig.CALENDAR_WIDGET_MONTH + widgetID);
            editor.remove(CalendarWidgetConfig.CALENDAR_WIDGET_YEAR + widgetID);
        }
        editor.apply();
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    @Override
    public void onAppWidgetOptionsChanged(Context ctxt, AppWidgetManager mgr, int appWidgetId, Bundle newOptions) {
        RemoteViews updateViews = new RemoteViews(ctxt.getPackageName(), R.layout.calendar_widget_layout);
        mgr.updateAppWidget(appWidgetId, updateViews);
    }
}

File: app/src/main/java/com/elementary/tasks/core/app_widgets/calendar/CalendarWidget.kt
package com.elementary.tasks.core.app_widgets.calendar

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.format.DateUtils
import android.widget.RemoteViews
import androidx.core.content.ContextCompat
import com.elementary.tasks.R
import com.elementary.tasks.core.app_widgets.WidgetUtils
import com.elementary.tasks.core.app_widgets.buttons.VoiceWidgetDialog
import com.elementary.tasks.core.os.PendingIntentWrapper
import com.elementary.tasks.home.BottomNavActivity
import com.elementary.tasks.reminder.create.CreateReminderActivity
import timber.log.Timber
import java.util.Calendar
import java.util.Formatter
import java.util.GregorianCalendar
import java.util.Locale

class CalendarWidget : AppWidgetProvider() {

  override fun onUpdate(
    context: Context,
    appWidgetManager: AppWidgetManager,
    appWidgetIds: IntArray
  ) {
    for (id in appWidgetIds) {
      updateWidget(context, appWidgetManager, CalendarWidgetPrefsProvider(context, id))
    }
    super.onUpdate(context, appWidgetManager, appWidgetIds)
  }

  override fun onAppWidgetOptionsChanged(
    context: Context,
    appWidgetManager: AppWidgetManager,
    appWidgetId: Int,
    newOptions: Bundle?
  ) {
    updateWidget(context, appWidgetManager, CalendarWidgetPrefsProvider(context, appWidgetId))
    super.onAppWidgetOptionsChanged(context, appWidgetManager, appWidgetId, newOptions)
  }

  companion object {

    fun updateWidget(
      context: Context,
      appWidgetManager: AppWidgetManager,
      sp: CalendarWidgetPrefsProvider
    ) {
      val options = appWidgetManager.getAppWidgetOptions(sp.widgetId)
      val width = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH)
      val height = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_HEIGHT)
      if (height != 0) {
        val rowHeight = (height - 58).toFloat() / 7f
        sp.setRowHeight(rowHeight)
        Timber.d("CALENDAR WIDGET SIZE w=$width, h=$height, row=$rowHeight")
      }

      val cal = GregorianCalendar()
      cal.set(Calendar.MONTH, sp.getMonth())
      cal.set(Calendar.YEAR, sp.getYear())
      val monthYearStringBuilder = StringBuilder(50)
      val monthYearFormatter = Formatter(
        monthYearStringBuilder, Locale.getDefault()
      )
      val monthYearFlag = (DateUtils.FORMAT_SHOW_DATE
        or DateUtils.FORMAT_NO_MONTH_DAY or DateUtils.FORMAT_SHOW_YEAR)
      val date = DateUtils.formatDateRange(
        context,
        monthYearFormatter, cal.timeInMillis, cal.timeInMillis, monthYearFlag
      ).toString().uppercase()

      val headerBgColor = sp.getHeaderBackground()
      val bgColor = sp.getBackground()

      val rv = RemoteViews(context.packageName, R.layout.widget_calendar)

      rv.setInt(R.id.headerBg, "setBackgroundResource", WidgetUtils.newWidgetBg(headerBgColor))
      rv.setInt(R.id.widgetBg, "setBackgroundResource", WidgetUtils.newWidgetBg(bgColor))

      rv.setTextViewText(R.id.widgetTitle, date)

      if (WidgetUtils.isDarkBg(headerBgColor)) {
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_settings_24px, R.color.pureWhite,
          R.id.btn_settings, CalendarWidgetConfigActivity::class.java
        ) {
          it.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
          return@initButton it
        }
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_add_24px, R.color.pureWhite,
          R.id.btn_add_task, CreateReminderActivity::class.java
        )
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_mic_24px, R.color.pureWhite,
          R.id.btn_voice, VoiceWidgetDialog::class.java
        )

        WidgetUtils.setIcon(
          context,
          rv,
          R.drawable.ic_twotone_keyboard_arrow_left_24px,
          R.id.btn_prev,
          R.color.pureWhite
        )
        WidgetUtils.setIcon(
          context,
          rv,
          R.drawable.ic_twotone_keyboard_arrow_right_24px,
          R.id.btn_next,
          R.color.pureWhite
        )

        rv.setTextColor(R.id.widgetTitle, ContextCompat.getColor(context, R.color.pureWhite))
      } else {
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_settings_24px, R.color.pureBlack,
          R.id.btn_settings, CalendarWidgetConfigActivity::class.java
        ) {
          it.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
          return@initButton it
        }
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_add_24px, R.color.pureBlack,
          R.id.btn_add_task, CreateReminderActivity::class.java
        )
        WidgetUtils.initButton(
          context, rv, R.drawable.ic_twotone_mic_24px, R.color.pureBlack,
          R.id.btn_voice, VoiceWidgetDialog::class.java
        )

        WidgetUtils.setIcon(
          context,
          rv,
          R.drawable.ic_twotone_keyboard_arrow_left_24px,
          R.id.btn_prev,
          R.color.pureBlack
        )
        WidgetUtils.setIcon(
          context,
          rv,
          R.drawable.ic_twotone_keyboard_arrow_right_24px,
          R.id.btn_next,
          R.color.pureBlack
        )

        rv.setTextColor(R.id.widgetTitle, ContextCompat.getColor(context, R.color.pureBlack))
      }

      val weekdayAdapter = Intent(context, CalendarWeekdayService::class.java)
      weekdayAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
      rv.setRemoteAdapter(R.id.weekdayGrid, weekdayAdapter)

      val startActivityIntent = Intent(context, BottomNavActivity::class.java)
      startActivityIntent.action = Intent.ACTION_VIEW
      val startActivityPendingIntent = PendingIntentWrapper.getActivity(
        context,
        0,
        startActivityIntent,
        PendingIntent.FLAG_MUTABLE,
        ignoreIn13 = true
      )
      rv.setPendingIntentTemplate(R.id.monthGrid, startActivityPendingIntent)

      val monthAdapter = Intent(context, CalendarMonthService::class.java)
      monthAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
      rv.setRemoteAdapter(R.id.monthGrid, monthAdapter)

      val nextIntent = Intent(context, CalendarNextReceiver::class.java)
      nextIntent.action = CalendarNextReceiver.ACTION_NEXT
      nextIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
      nextIntent.putExtra(CalendarNextReceiver.ARG_VALUE, 2)
      val nextPendingIntent =
        PendingIntentWrapper.getBroadcast(context, 0, nextIntent, PendingIntent.FLAG_UPDATE_CURRENT)
      rv.setOnClickPendingIntent(R.id.btn_next, nextPendingIntent)

      val previousIntent = Intent(context, CalendarPreviousReceiver::class.java)
      previousIntent.action = CalendarPreviousReceiver.ACTION_PREVIOUS
      previousIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, sp.widgetId)
      previousIntent.putExtra(CalendarPreviousReceiver.ARG_VALUE, 1)
      val previousPendingIntent =
        PendingIntentWrapper.getBroadcast(
          context,
          0,
          previousIntent,
          PendingIntent.FLAG_UPDATE_CURRENT
        )
      rv.setOnClickPendingIntent(R.id.btn_prev, previousPendingIntent)

      appWidgetManager.updateAppWidget(sp.widgetId, rv)
      appWidgetManager.notifyAppWidgetViewDataChanged(sp.widgetId, R.id.weekdayGrid)
      appWidgetManager.notifyAppWidgetViewDataChanged(sp.widgetId, R.id.monthGrid)
    }
  }
}
